### ISSUE TITLE
[tt-train] grad sync and clipping should happen only during stepping in the presence of grad accumulation

### ISSUE THREAD

**Issue by jaykru-tt** [tenstorrent/tt-metal#17568]
**Describe the bug**
In the nanogpt demo, we currently clip grads after every backward pass, regardless of whether the gradient accumulation helper says it's time to step. We need to change the demo so that grads are clipped only before an optimizer step is to take place.

**Comment by zzigler-tt** [comment]
@rfurko-tt can you please review and determine if this still an issue or can be closed out. Thanks

**Comment by zzigler-tt** [comment]
@ayerofieiev-tt can you please review and determine if this still an issue or can be closed out. Thanks
